### PR TITLE
Gh 1842/delegate warning flag

### PR DESCRIPTION
**Pull request by biafra23**
Handles #1842 

Changes proposed in this pull request:

TX Details | Action List | Action details
---------- | ----------- | --------------
image | image | image

### CODE REVIEW

File: Multisig/Data/Services/Utils/DelegateWarningCalculator.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+// used to enhance txData to contain trustedDelegateCallTarget flags in every displayed tx (also multisend inner txs)
+class DelegateWarningCalculator {
+    static func addMissingTrustedDelegateCallTargets(txData: inout SCGModels.TxData) {
+
+        // TODO check: if operation is DELEGATE then to-address must be in addressInfoIndex otherwise trustedDelegateCallTarget = false
+        // if txData.to
+
+        // for testing. This should replace any false value with true ad thus fail the first test
+        txData.trustedDelegateCallTarget = true
```
Comment (biafra23) [inline Multisig/Data/Services/Utils/DelegateWarningCalculator.swift]: This is supposed to modify the parameter (`txData`) but as you can see in the test result, ist doesn't. Same for the other modifications later in this method.

File: MultisigTests/Logic/Models/Transaction/DelegateWarningCalculatorTests.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    func testTopLevelWarning() throws {
+        // https://safe-client.gnosis.io/v1/chains/4/transactions/multisig_0x73a7AA145338587f7aB7f63c06d187C85dF4727e_0xba28109747222d6cfb7f0fb75f03d49957e343674bc116162b038e244dbcc6d6
+        let data = jsonData("DelegateWarningTopLevel")
+        let decoder = JSONDecoder()
+        decoder.dateDecodingStrategy = .millisecondsSince1970
+        var decodedData = try decoder.decode(SCGModels.TransactionDetails.self, from: data)
+
+        DelegateWarningCalculator.addMissingTrustedDelegateCallTargets(txData: &decodedData.txData!)
```
Comment (DmitryBespalov) [inline MultisigTests/Logic/Models/Transaction/DelegateWarningCalculatorTests.swift]: can you try to put `txData` in another variable before calling this function?

Comment (DmitryBespalov) [reply MultisigTests/Logic/Models/Transaction/DelegateWarningCalculatorTests.swift]: maybe because it is part of another variable it is not getting modified...

Comment (biafra23) [reply MultisigTests/Logic/Models/Transaction/DelegateWarningCalculatorTests.swift]: Does not seem to help. See my recent changes. 
I introduced variable and modified the test. When the parameter is modified successfully, now the test should stop failing.

File: Multisig/Data/Services/Utils/DelegateWarningCalculator.swift (hunk 1 above)
Comment (DmitryBespalov) [reply Multisig/Data/Services/Utils/DelegateWarningCalculator.swift]: 1) this is modified OK, i.e. before this line, the variable is false

File: Multisig/Data/Services/Utils/DelegateWarningCalculator.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                if case let SCGModels.DataDecoded.Parameter.ValueDecoded.multiSend(multiSendTxs)? = parameter.valueDecoded {
+                    print("--> Multisend: (\(multiSendTxs.count) actions)")
+                    for var multiSendTx in multiSendTxs {
+                        if multiSendTx.operation == .delegate {
+                            print("--> Multisend: is DELEGATE, addressInfo: \(txData.addressInfoIndex?.values[multiSendTx.to]?.addressInfo)")
+                            if txData.addressInfoIndex?.values[multiSendTx.to]?.addressInfo == nil {
+                                print("--> Multisend: multiSendTx.trustedDelegateCallTarget = false")
+                                multiSendTx.trustedDelegateCallTarget = false
+                                // TODO: also set surrounding tx trustedDelegateCallTarget to false
+                                txData.trustedDelegateCallTarget = false
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Utils/DelegateWarningCalculator.swift]: 2) after debugging, I see that this line is executed, which sets the variable to `false`.

File: MultisigTests/Logic/Models/Transaction/DelegateWarningCalculatorTests.swift [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        let decoder = JSONDecoder()
+        decoder.dateDecodingStrategy = .millisecondsSince1970
+        var decodedData = try decoder.decode(SCGModels.TransactionDetails.self, from: data)
+
+        var txData: SCGModels.TxData = decodedData.txData!
+        XCTAssertFalse(txData.trustedDelegateCallTarget!)
+
+        DelegateWarningCalculator.addMissingTrustedDelegateCallTargets(txData: &txData)
+
+        XCTAssertTrue(txData.trustedDelegateCallTarget!)
```
Comment (DmitryBespalov) [inline MultisigTests/Logic/Models/Transaction/DelegateWarningCalculatorTests.swift]: 3) here, the variable is `false`

File: Multisig/Data/Services/Utils/DelegateWarningCalculator.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                            } else {
+                                print("--> Multisend: multiSendTx.trustedDelegateCallTarget = true")
+                                multiSendTx.trustedDelegateCallTarget = true
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Utils/DelegateWarningCalculator.swift]: by the way, the modifications here are made to the in-memory variable multiSendTx which is a copy of the data inside the txData

File: Multisig/Data/Services/Utils/DelegateWarningCalculator.swift [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        txData.trustedDelegateCallTarget = true
+
+        if txData.dataDecoded?.method == "multiSend" {
+            guard let parameters = txData.dataDecoded?.parameters else {
+                return
+            }
+            for var parameter in parameters {
+                if case let SCGModels.DataDecoded.Parameter.ValueDecoded.multiSend(multiSendTxs)? = parameter.valueDecoded {
+                    print("--> Multisend: (\(multiSendTxs.count) actions)")
+                    for var multiSendTx in multiSendTxs {
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Utils/DelegateWarningCalculator.swift]: in each loop iteration, there will be a fresh copy of the multiSendTx

Comment (biafra23) [reply Multisig/Data/Services/Utils/DelegateWarningCalculator.swift]: How can I access the `MultiSendTx`s without creating a copy?

Comment (DmitryBespalov) [reply Multisig/Data/Services/Utils/DelegateWarningCalculator.swift]: you would need to access it via subscript, if it would be possible. The problem is with enumerations.

Ideally it would be possible to do something like

```
txData.dataDecoded.parameters[i].valueDecoded[j].trustedDelegateCallTarget = false
```

But I don't see Swift allowing this to do.

Alternative is to construct new parameters from the old ones with only different `trustedDelegateCallTarget` value

**Comment by DmitryBespalov** [comment]
@Dirk I think the vertical padding is too big compared to the designs
padding-delegate

**Comment by DmitryBespalov** [comment]
As well as the distance from the "( i )" icon to the text

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift [hunk 7]
```diff
@@ -503,6 +503,7 @@ extension SCGModels {
         var hexData: DataString?
         var dataDecoded: DataDecoded?
         var addressInfoIndex: AddressInfoIndex?
+        var trustedDelegateCallTarget: Bool?
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift:506]: I think this can be removed

File: Multisig/Data/Services/Utils/DelegateWarningCalculator.swift [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+        let result = allParametersTrusted ? TRUSTED : UNTRUSTED
+
+        return result
+    }
+
+    /// Multi Send transaction is potentially dangerous (untrusted) if it is a "delegate call"
+    /// operation and the recipient (target) address, "to", does not have a name resolved.
+    ///
+    /// If the multisend transaction is itself a multi-send
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Utils/DelegateWarningCalculator.swift]: this line seems to be irrelevant now

File: Multisig/UI/Transaction/TransactionDetailsViewController/TransactionDetailCellBuilder.swift [hunk 9]
```diff
@@ -659,14 +669,19 @@ class TransactionDetailCellBuilder {
                  title: String?,
                  imageUri: URL? = nil,
                  browseURL: URL? = nil,
-                 prefix: String? = nil) {
+                 prefix: String? = nil,
+                 showDelegateWarning: Bool = false) {
+
+        print("\(address.ellipsized()) showDelegateWarning: \(showDelegateWarning)")
```
Comment (DmitryBespalov) [inline Multisig/UI/Transaction/TransactionDetailsViewController/TransactionDetailCellBuilder.swift]: debug print can be removed

**Review by DmitryBespalov — COMMENTED** [review]
Epic!

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift (hunk 7 above)
Comment (biafra23) [reply Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift:506]: I think this one is necessary. It is the flag directly inside TxData, isn't it?

File: Multisig/UI/Transaction/TransactionDetailsViewController/TransactionDetailCellBuilder.swift (hunk 9 above)
Comment (biafra23) [reply Multisig/UI/Transaction/TransactionDetailsViewController/TransactionDetailCellBuilder.swift]: upsi

**Comment by biafra23** [comment]
@DmitryBespalov I wanted to postpone the link to the next PR. When the article is available.

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift (hunk 7 above)
Comment (DmitryBespalov) [reply Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift:506]: aah, true!

**Comment by biafra23** [comment]
Link has been added. But I had to allow changes to the font size to make all of it fit. See screenshots

**Comment by DmitryBespalov** [comment]
Still, the height of the warning seems to big, i.e. the vertical distance from text to the top and bottom edges

**Comment by biafra23** [comment]
> Still, the height of the warning seems to big, i.e. the vertical distance from text to the top and bottom edges

Now I get it. The comment wasn't about the missing link.

**Review by DmitryBespalov — APPROVED** [review]
Looks great! 🚀
